File: Problems/Algorithms/1396. Design Underground System/UndergroundSystem.java
class UndergroundSystem {
    private Map<Integer, Pair<String, Integer>> checkIns;
    private Map<Pair<String, String>, int[]> times;

    public UndergroundSystem() {
        checkIns = new HashMap<>();
        times = new HashMap<>();
    }
    
    public void checkIn(int id, String stationName, int t) {
        checkIns.put(id, new Pair(stationName, t));
    }
    
    public void checkOut(int id, String stationName, int t) {
        String startStation = checkIns.get(id).getKey();
        int startTime = checkIns.get(id).getValue();
        checkIns.remove(id);

        Pair<String, String> pair = new Pair(startStation, stationName);
        int totalTime = times.containsKey(pair) ? times.get(pair)[0] : 0;
        int number = times.containsKey(pair) ? times.get(pair)[1] : 0;
        times.put(pair, new int[] {totalTime + t - startTime, number + 1});
    }
    
    public double getAverageTime(String startStation, String endStation) {
        Pair<String, String> pair = new Pair(startStation, endStation);

        return (double) times.get(pair)[0] / times.get(pair)[1];
    }
}

/**
 * Your UndergroundSystem object will be instantiated and called as such:
 * UndergroundSystem obj = new UndergroundSystem();
 * obj.checkIn(id,stationName,t);
 * obj.checkOut(id,stationName,t);
 * double param_3 = obj.getAverageTime(startStation,endStation);
 */

File: Problems/Algorithms/1396. Design Underground System/UndergroundSystem.kt
class UndergroundSystem() {
    
    private val checkins: MutableMap<Int, Pair<String, Int>> = mutableMapOf()
    private val trajectories: MutableMap<String, Pair<Int, Int>> = mutableMapOf()

    fun checkIn(id: Int, stationName: String, t: Int) {
        checkins[id] = stationName to t
    }

    fun checkOut(id: Int, stationName: String, t: Int) {
        val from = checkins[id]!!
        val traj = from.first + "_" + stationName
        val duration = t - from.second
        
        val previous = trajectories[traj] ?: 0 to 0
        trajectories[traj] = previous.first + duration to previous.second + 1
    }

    fun getAverageTime(startStation: String, endStation: String): Double {
        val pair = trajectories[startStation + "_" + endStation] ?: 0 to 0
        return pair.first.toDouble() / pair.second.toDouble()
    }

}

/**
 * Your UndergroundSystem object will be instantiated and called as such:
 * var obj = UndergroundSystem()
 * obj.checkIn(id,stationName,t)
 * obj.checkOut(id,stationName,t)
 * var param_3 = obj.getAverageTime(startStation,endStation)
 */
